### ISSUE TITLE
Bug in modeling_bart.eager_attention_forward

### ISSUE THREAD

**Issue by xadupre** [huggingface/transformers#39365]
### System Info

In [sdpa_forward_attention](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/sdpa_attention.py#L44),
we have this:

```python
if attention_mask is not None and attention_mask.ndim == 4:
    attention_mask = attention_mask[:, :, :, : key.shape[-2]]
```

In [eager_attention_foward](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bart/modeling_bart.py#L133), these two lines are not there and it may fail there:

```python
    if attention_mask is not None:
        attn_weights = attn_weights + attention_mask
```

I was wondering if that was intended. sdpa seems to be the default implementation.

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Switch default attn_implementation to eager.

### Expected behavior

Both implementation should work.

**Comment by vasqu** [comment]
Do you have a small reproducer that fails? But, this seems valid - would you open a PR to fix this (ideally with a small test to catch this)?

**Comment by zucchini-nlp** [comment]
Ideally, would be great if we can copy all `eager_attention_foward` functions from llama, keeping only one source of truth

**Comment by vasqu** [comment]
Ideally, yes but there were a few differences (fp32 softmax) and legacy features (head mask), so it made more sense to keep them separate for now. I agree tho that we should keep the number of real `eager_attention_forward`s to a minimum 👀

**Comment by xadupre** [comment]
This bug happens for the model facebook/bart-base. I created a unit test which replicates the issue: https://github.com/huggingface/transformers/pull/39435. The inputs I chose may be not expected but by the model but they are working with the sdpa implementation.

**Comment by vasqu** [comment]
Not stale, keeping tabs on it

**Comment by qgallouedec** [comment]
Hey @xadupre, sorry for the slow reply, we're catching up on the backlog. This is fixed on main.

Since #41265 the Bart decoder builds its masks through `masking_utils`, so the mask handed to the attention interface is derived from the actual key/value shapes instead of the raw 2D mask:

https://github.com/huggingface/transformers/blob/ff4bb5ee67be830a9bd45426ca5626825c3f222a/src/transformers/models/bart/modeling_bart.py#L640-L645

Your reproducer from #39435 now runs under both `eager` and `sdpa`, with logits agreeing to ~7e-9. The slice in `sdpa_attention_forward` that motivated the asymmetry was itself removed in #41900, so both paths now receive the identical mask. Closing as fixed, but please re-open if you still hit this on a recent version.
